docker.md

```diff
@@ -0,0 +1,122 @@
+
+If you need to add any environment variables, you can do so in the `docker-compose.yml`
+files as shown, or add them to `.env` and load them in as shown
+[here](https://docs.docker.com/compose/compose-file/#env_file).
+
+### Stages
+There are currently four stages in the Dockerfile that are used.
+
+#### `base`
+Contains all package installs, files system modifications, and entrypoint declaration.
+Any commands that will not change across the other stages or are required for multiple
+other stages to execute properly should be put in this stage.
+
+#### `dependencies`
+The `dependencies` stage is used to install npm dependencies, copy all of your source
+files, and then optionally run any build commands that need to be executed. The desired
+outcome of this stage is an image that contains all dependencies, all source files, and
+all built assets (if necessary).
+
+In order to run build scripts, you will need to locate this stage in the Dockerfile
+and uncomment the `RUN npm run build` command that is in there, possibly changing the
+command if you are using multiple build scripts in your `package.json`.
+
+#### `test`
+The `test` stage is used for running your test suite. It changes the user to `node`
+to mimic the production environment, and then runs `npm lint && npm test` which should
+run your full test suite. The container will exit with the same code as your test
+runner, which means that in CI you can run `npm run docker:test:ci` as your test step to
+automatically build your test image and then run your tests inside of it.
+
+#### `release`
+The `release` stage is meant to run your server in a production-like setup. The
+`docker-compose.yml` file can be used to tweak the environment variables used for
+this stage when running locally, and the logic for building this stage should not
+make any assumptions about what environment the output will be used in.
+
+## Local Development
+If you are using the `development` flavor alongside this `docker` flavor, you can simply
+run `npm run docker:start` to build the release stage of the Dockerfile, and then start
+it as a `daemon` in Docker. You can then visit `localhost:3000` to hit your server.
+
+By default, the server is started using `nodemon` when you start it using
+`docker-compose`, which means every time you save a file the server will be restarted
+automatically with your fresh changes.
+
+### Running Tests
+Running tests locally is extremely easy. Simply run `npm run docker:test`, and linting
+and testing will be run. If you modify the `package*.json` files, you will need to run
+`npm run build:test` to trigger a reinstall of your dependencies.
+
+### Secret Management (`.env`)
+This removes the option to have a `.env` files for secrets locally, because those environment
+variables can now just be set in the `docker-compose.yml` file `environment` section. If
+you have true secrets that you need locally and cannot check into Git, you can re-create
```

Comment:
Our .env setup is really for this purposes: secrets/credentials that you need locally and can't check-in.  I would love to still be able to speak to that in some way.  Any ideas?

Reply:
Do you find

>Secret Management (.env)
>This removes the option to have a .env files for secrets locally, because those environment variables can now just be set in the docker-compose.yml file environment section. If you have true secrets that you need locally and cannot check into Git, you can re-create the server/.env file, and then add a key in the *_base services in docker-compose.yml that is simply env_file: server/.env at the same level in the yaml as the environment key.

to be lacking in information? it and a section farther up the file talk about adding the `env_file` key to `docker-compose` if you need to use the local file

Reply:
Ah no, that makes perfect sense!  I would just prefer to have something setup in the flavor prepared to work "out of the box" if the user opts to go that direction.  For example, if the user follows this advice then they'll run into the same issues that we're currently working through e.g. checking-in the .env but not tracking it, etc.  I'm hoping to have a good, default answer.

Based on our conversations in slack I see a few options, but they all seem to come with caveats.  I still think we can work something out, between `--env-file`, `-f server/docker-compose.yml`, `docker-compose.override.yml`, checking-in the .env file and ignoring it, etc.

Reply:
:+1: i can restore the `.env.example` file and just comment out the `docker-compose.yml` lines pointing to `server/.env` so that all that is required is uncommenting the lines and creating the `.env` file if that would be preferred.

Reply:
Alright, let's do that.  I think the only leftover question is whether to move `.env-keep` to the project root or to adjust the commands to work with `server/.env`.

Reply:
I think adjusting is easiest - fewer moving parts between flavors. I'll work on this when I get back from Chicago, hopefully tomorrow or Friday.

Reply:
I think it would be easier at the project root.
But that means the file should at same place for every flavors of the boilerplate too, to be consistent.